mmcv/ops/csrc/common/cuda/correlation_cuda.cuh

```diff
@@ -0,0 +1,269 @@
+
+#ifdef MMCV_USE_PARROTS
+#include "parrots_cuda_helper.hpp"
+#else
+#include "pytorch_cuda_helper.hpp"
+#endif
+
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <torch/types.h>
+#include <vector>
+#include <iostream>
+
+using namespace torch;
+
+#define TensorAcc4R PackedTensorAccessor32<scalar_t, 4, RestrictPtrTraits>
+#define TensorAcc5R PackedTensorAccessor32<scalar_t, 5, RestrictPtrTraits>
+#define WITHIN_BOUNDS(x, y, H, W) (x >= 0 && x < H && y >= 0 && y < W)
+
+#define THREADS_FORWARD 32
+#define THREADS_BACKWARD 16
+
+template <typename scalar_t>
+__global__ void correlation_forward_cuda_kernel(const TensorAcc4R rInput1,
+                                                const TensorAcc4R rInput2,
+                                                TensorAcc5R output,
+                                                int kH, int kW,
+                                                int patchH, int patchW,
+                                                int padH, int padW,
+                                                int dilationH, int dilationW,
+                                                int dilation_patchH,
+                                                int dilation_patchW,
+                                                int dH, int dW)
+{
+  const int iH = rInput1.size(1);
+  const int iW = rInput1.size(2);
+  const int C = rInput1.size(3);
+
+  const int n = blockIdx.x;
+  const int h = blockIdx.y;
+  const int w = blockIdx.z;
+  const int thread = threadIdx.x;
+
+  const int start_i = -padH + h * dH;
+  const int start_j = -padW + w * dW;
+
+  const int patchRadH = dilation_patchH * (patchH - 1) / 2;
+  const int patchRadW = dilation_patchW * (patchW - 1) / 2;
+
+  __shared__ scalar_t prod_sum[THREADS_FORWARD];
+
+  for (int ph = 0; ph < patchH; ++ph)
+  {
+    int ph_dilated = ph * dilation_patchH - patchRadH;
+    for (int pw = 0; pw < patchW; ++pw)
+    {
+      int pw_dilated = pw * dilation_patchW - patchRadW;
+      prod_sum[thread] = 0;
+      for (int i = 0; i < kH; ++i)
+      {
+        int i1 = start_i + i * dilationH;
+        int i2 = i1 + ph_dilated;
+        if WITHIN_BOUNDS (i1, i2, iH, iH)
+        {
+          for (int j = 0; j < kW; ++j)
+          {
+            int j1 = start_j + j * dilationW;
+            int j2 = j1 + pw_dilated;
+            if WITHIN_BOUNDS (j1, j2, iW, iW)
+            {
+              for (int c = thread; c < C; c += THREADS_FORWARD)
+              {
+                scalar_t v1 = rInput1[n][i1][j1][c];
+                scalar_t v2 = rInput2[n][i2][j2][c];
+                prod_sum[thread] += v1 * v2;
+              }
+            }
+          }
+        }
+      }
+      // accumulate
+      __syncthreads();
+      if (thread == 0)
+      {
+        scalar_t reduce_sum = 0;
+        for (int index = 0; index < THREADS_FORWARD; ++index)
+        {
+          reduce_sum += prod_sum[index];
+        }
+        output[n][ph][pw][h][w] = reduce_sum;
+      }
```

Comment:
Since you have a fix `THREADS_FORWARD==32`, you can use [Warp-Level Primitives](https://developer.nvidia.com/blog/using-cuda-warp-level-primitives/) to do the reduction sum. You can even use a register per thread instead of shared memory to store the prod_sum.
Just advice, you can still keep the code unchanged.